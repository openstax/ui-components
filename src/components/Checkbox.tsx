import { PropsWithChildren } from "react";
import { colors } from "../theme";
import styled from "styled-components";
import { InputHTMLAttributes } from "react";

type CheckboxVariant = keyof typeof checkboxVariants;
type CheckboxSize = 1.4 | 1.6 | 1.8 | 2;

export const checkboxVariants = {
  primary: {
    accentColor: colors.palette.mediumBlue,
    boxShadow: 'none',
  },
  light: {
    accentColor: colors.palette.white,
    boxShadow: '0 0 1px 0',
  }
} as const;

const StyledLabel = styled.label<{ bold: boolean; }>`
  font-size: 1.4rem;
  display: flex;
  align-items: center;
  font-weight: ${props => props.bold ? 700 : 400}
`;

const StyledInput = styled.input<{ variant: CheckboxVariant; checkboxSize: CheckboxSize; }>`
  accent-color: ${props => checkboxVariants[props.variant].accentColor};
  width: ${props => props.checkboxSize}rem;
  height: ${props => props.checkboxSize}rem;
  margin: 0 1.6rem 0 0;
  &:checked {
    box-shadow: ${props => checkboxVariants[props.variant].boxShadow};
  }
`;

type CheckboxProps = PropsWithChildren<
  Omit<InputHTMLAttributes<HTMLInputElement>, 'type'> & {
  variant?: CheckboxVariant;
  size?: CheckboxSize;
  bold?: boolean;
}>;

export const Checkbox = ({ children, variant = 'primary', bold = false, size = 1.6, ...props }: CheckboxProps) => {
  return (
    <StyledLabel bold={bold}>
      <StyledInput {...props} type="checkbox" variant={variant} checkboxSize={size} />
      {children}
    </StyledLabel>
  );
};
