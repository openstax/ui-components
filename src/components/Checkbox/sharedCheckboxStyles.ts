import { colors } from "../../theme";
import { whiteCheckmark, grayCheckmark, redCheckmark } from "../svgs/checkmarksvgs";

export type CheckboxVariant = keyof typeof checkboxVariants;
export type CheckboxSize = 1.4 | 1.6 | 1.8 | 2;

export const checkboxVariants = {
  primary: {
    backgroundColor: colors.palette.mediumBlue,
    color: 'inherit',
    unCheckedBorder: `1px solid ${colors.palette.neutralThin}`,
    checkedBorder: `1px solid ${colors.palette.mediumBlue}`,
    backgroundImage: whiteCheckmark
  },
  light: {
    backgroundColor: colors.palette.white,
    color: 'inherit',
    unCheckedBorder: `1px solid ${colors.palette.pale}`,
    checkedBorder: `1px solid ${colors.palette.pale}`,
    backgroundImage: grayCheckmark
  },
  error: {
    backgroundColor: colors.palette.paleRed,
    color: colors.palette.darkRed,
    unCheckedBorder: `1px solid ${colors.palette.lightRed}`,
    checkedBorder: `1px solid ${colors.palette.lightRed}`,
    backgroundImage: redCheckmark
  },
  disabled: {
    backgroundColor: colors.palette.white,
    color: 'inherit',
    unCheckedBorder: `1px solid ${colors.palette.pale}`,
    checkedBorder: `1px solid ${colors.palette.pale}`,
    backgroundImage: 'none'
  }
} as const;