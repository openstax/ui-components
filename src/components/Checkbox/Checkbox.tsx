import { LabelHTMLAttributes, PropsWithChildren } from "react";
import {  checkboxLabelStyles, checkboxInputStyles, CheckboxVariant, CheckboxSize } from "./sharedCheckboxStyles";
import styled from "styled-components";
import { InputHTMLAttributes } from "react";

const StyledLabel = styled.label<{ bold: boolean; variant: CheckboxVariant; isDisabled?: boolean; }>`
  ${checkboxLabelStyles}
`;

// https://moderncss.dev/pure-css-custom-checkbox-style/
const StyledInput = styled.input<{ variant: CheckboxVariant; checkboxSize: CheckboxSize; isDisabled?: boolean; }>`
  ${checkboxInputStyles}
`;

type CheckboxProps = PropsWithChildren<
  Omit<InputHTMLAttributes<HTMLInputElement>, 'type'> & {
  variant?: CheckboxVariant;
  size?: CheckboxSize;
  bold?: boolean;
  labelProps?: LabelHTMLAttributes<HTMLLabelElement>;
}>;

export const Checkbox = ({ children, disabled, variant = 'primary', bold = false, size = 1.6, labelProps, ...props }: CheckboxProps) => {
  return (
    <StyledLabel bold={bold} variant={variant} isDisabled={disabled} {...labelProps}>
      <StyledInput {...props} type="checkbox" variant={variant} checkboxSize={size} isDisabled={disabled} />
      {children}
    </StyledLabel>
  );
};
