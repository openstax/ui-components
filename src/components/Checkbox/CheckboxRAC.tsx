import { Checkbox as RACCheckbox, CheckboxProps as RACCheckboxProps } from "react-aria-components";
import styled from "styled-components";
import { PropsWithChildren } from "react";
import { checkboxVariants } from "./sharedCheckboxStyles";
import { colors } from "../../theme";

type Variant = keyof typeof checkboxVariants;
type Size = 1.4 | 1.6 | 1.8 | 2;

const StyledCheckbox = styled(RACCheckbox)<{
  $variant: Variant;
  $size: Size;
  $bold: boolean;
}>`
  display: flex;
  align-items: center;
  font-size: 1.6rem;
  font-weight: ${({ $bold }) => ($bold ? 700 : 400)};
  gap: 1.2rem;
  color: inherit;

  &[data-checked] [data-slot="selection"] {
    background-color: ${({ $variant }) => checkboxVariants[$variant].backgroundColor};
    background-image: url(${({ $variant }) => checkboxVariants[$variant].backgroundImage});
    background-position: center;
    background-repeat: no-repeat;
    background-size: 80%;
    border: ${({ $variant }) => checkboxVariants[$variant].checkedBorder};
  }

  [data-slot="selection"] {
    width: ${({ $size }) => $size}rem;
    height: ${({ $size }) => $size}rem;
    margin: 0;
    border-radius: 0.2rem;
    background-color: ${colors.palette.white};
    border: ${({ $variant }) => checkboxVariants[$variant].unCheckedBorder};
    transition: all 0.2s ease-in-out;
  }

  &[data-disabled] [data-slot="selection"] {
    opacity: 0.4;
    cursor: not-allowed;
  }
`;

export interface CheckboxProps extends PropsWithChildren<Omit<RACCheckboxProps, 'children'>> {
  size?: Size;
  variant?: Variant;
  bold?: boolean;
}

export const CheckboxRAC = ({
  size = 1.6,
  variant = 'primary',
  bold = false,
  children,
  ...props
}: CheckboxProps) => {
  return (
    <StyledCheckbox $size={size} $variant={variant} $bold={bold} {...props}>
      {children}
    </StyledCheckbox>
  );
};