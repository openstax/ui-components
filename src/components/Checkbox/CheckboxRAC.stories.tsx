import styled from "styled-components";
import { CheckboxRAC } from "./CheckboxRAC";

const CheckboxGroup = styled.div`
  text-transform: capitalize;

  & + & {
    margin-top: 3.2rem;
  }

  > * + * {
    margin-top: 0.5rem;
  }
`;

type CheckboxProps = React.ComponentProps<typeof CheckboxRAC>;

const renderCheckboxes = (
  variant: CheckboxProps['variant'],
  size: CheckboxProps['size']
) => (
  <CheckboxGroup>
    <h2>Size {size}</h2>
    <CheckboxRAC slot='selection' {...{ size, variant }}>Checkbox label</CheckboxRAC>
    <CheckboxRAC slot='selection' {...{ size, variant }} defaultSelected>
      Checkbox label
    </CheckboxRAC>
    <CheckboxRAC slot='selection' {...{ size, variant }} defaultSelected bold>
      Checkbox label
    </CheckboxRAC>
  </CheckboxGroup>
);

export const Primary = () => (
  <>
    {renderCheckboxes("primary", 1.4)}
    {renderCheckboxes("primary", 1.6)}
    {renderCheckboxes("primary", 1.8)}
    {renderCheckboxes("primary", 2)}
  </>
);

export const Light = () => (
  <>
    {renderCheckboxes("light", 1.4)}
    {renderCheckboxes("light", 1.6)}
    {renderCheckboxes("light", 1.8)}
    {renderCheckboxes("light", 2)}
  </>
);

export const Error = () => (
  <>
    {renderCheckboxes("error", 1.4)}
    {renderCheckboxes("error", 1.6)}
    {renderCheckboxes("error", 1.8)}
    {renderCheckboxes("error", 2)}
  </>
);

export const Disabled = () => (
  <>
    <CheckboxGroup>
      <h2>Disabled</h2>
      <CheckboxRAC slot='selection' variant="primary" size={1.6} isDisabled>
        Checkbox label
      </CheckboxRAC>
      <CheckboxRAC slot='selection' variant="primary" size={1.6} isDisabled defaultSelected>
        Checkbox label
      </CheckboxRAC>
    </CheckboxGroup>
  </>
);