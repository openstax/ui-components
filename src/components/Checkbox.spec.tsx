import { Checkbox } from './Checkbox';
import renderer from 'react-test-renderer';

describe('Checkbox', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <Checkbox>Click Me</Checkbox>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles options', () => {
    const tree = renderer.create(
      <Checkbox bold size={2} variant='light'>Click Me</Checkbox>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('allows setting props on label', () => {
    const tree = renderer.create(
      <Checkbox bold size={2} variant='light' labelProps={{ "aria-label": "custom label"}}>
        Click Me
      </Checkbox>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('handles disabled state', () => {
    const tree = renderer.create(
      <Checkbox disabled>Click Me</Checkbox>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
