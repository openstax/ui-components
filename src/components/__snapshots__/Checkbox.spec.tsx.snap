// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox handles options 1`] = `
<label
  className="sc-bczRLJ bYTfrb"
>
  <input
    className="sc-gsnTZi hvmWuC"
    type="checkbox"
  />
  Click Me
</label>
`;

exports[`Checkbox matches snapshot 1`] = `
<label
  className="sc-bczRLJ bYTeUU"
>
  <input
    className="sc-gsnTZi ljReaQ"
    type="checkbox"
  />
  Click Me
</label>
`;
